tap-agent/src/agent/sender_allocation.rs

```diff
@@ -476,6 +476,45 @@ impl SenderAllocationState {
                 ),
                 _ => e.into(),
             })?;
+
+        // store invalid receipts and if all are invalid remove them
+        if !invalid_receipts.is_empty() {
+            warn!(
+                "Found {} invalid receipts for allocation {} and sender {}.",
+                invalid_receipts.len(),
+                self.allocation_id,
+                self.sender
+            );
+
+            // Save invalid receipts to the database for logs.
+            // TODO: consider doing that in a spawned task?
+            self.store_invalid_receipts(invalid_receipts.as_slice())
+                .await?;
+            if valid_receipts.is_empty() {
+                let min_timestamp = invalid_receipts
+                    .first()
+                    .expect("There should be atleast one receipt")
+                    .signed_receipt()
+                    .message
+                    .timestamp_ns;
+                let max_timestamp = invalid_receipts
+                    .last()
+                    .expect("There should be atleast one receipt")
+                    .signed_receipt()
+                    .message
+                    .timestamp_ns;
+                sqlx::query!(
+                    r#"
+                        DELETE FROM scalar_tap_receipts
+                        WHERE timestamp_ns BETWEEN $1 AND $2;
+                    "#,
+                    BigDecimal::from(min_timestamp),
+                    BigDecimal::from(max_timestamp)
+                )
```

Comment:
This is not gonna work correctly I think.

Right now this is where obsolete receipts are deleted: https://github.com/graphprotocol/indexer-rs/blob/d8d1af901cc6ce4dbc65303e7c4ae205c41d8dee/tap-agent/src/agent/sender_allocation.rs#L322. The gist of it is that from time to time, we look at the latest RAV we got in the DB (which must be valid), and delete all receipts that are older than that.
In the case where we got a successful RAV and there were a few failed receipts that were not included, it's fine because the failed receipts, even though not aggregated, will be deleted if they're older than the last RAV.

So the fix we really need is only for the special case of the RAV request failing because *none* of the receipts were valid in the batch. In that case we're stuck trying over and over again because there is no newer RAV that will make it so that eventually the batch of bad receipts gets deleted.
So we'd only look for [this error](https://github.com/semiotic-ai/timeline-aggregation-protocol/blob/eb8447ed4566ced6846c03b510b25b915f985186/tap_core/src/manager/tap_manager.rs#L212) or just `valid_receipts` being empty, and handle the special case there.

We have to be careful with the `BETWEEN` to make sure that it is inclusive both on the lower and higher bound (though please double check it's what we indeed need here).

Also we may consider adding an adapter in tap_core for the removal of receipts in a range, just to make more uniform.

Reply:
Ok so I was wrong here, I was confused and missed the `if valid_receipts.is_empty()`. The logic is OK but perhaps there is a way to make it easier to see.

Reply:
yeah, switched  to use a match using `expected_rav` and added also `invalid` and `valid` receipts into a tuple as sugested